File: typescript/coding_agent/src/workflow.ts
/**
 * Coding agent workflow.
 *
 * Orchestrates the full test-driven development loop:
 *   1. Plan — generate dev_plan + test_plan
 *   2. Iterate (up to max_retries times):
 *      a. Iteration 0: generate tests first, then code (tests-visible codegen)
 *      b. Iteration N>0: analyze errors, then fix code
 *      c. Sync code + tests to E2B sandbox
 *      d. Install third-party dependencies
 *      e. Run pytest
 *      f. Decision: success → docs → return | failure → next iteration
 */

import { workflow } from '@agnt5/sdk';
import type { Context } from '@agnt5/sdk';

import type { WorkflowResult, ErrorAnalysis } from './models.js';
import {
  plannerNode,
  codeGeneratorNode,
  testGeneratorNode,
  codeSyncNode,
  installDepsNode,
  codeExecutorNode,
  errorAnalyzerNode,
  finalResponseNode,
} from './functions.js';

export const codingAgentWorkflow = workflow(
  'coding_agent_workflow',
  async (
    ctx: Context,
    input: { task_description: string; max_retries?: number },
  ): Promise<WorkflowResult> => {
    const { task_description, max_retries = 15 } = input;

    ctx.logger.info('='.repeat(60));
    ctx.logger.info('INITIATING CODING AGENT WORKFLOW');
    ctx.logger.info('='.repeat(60));

    await ctx.set('task_description', task_description);
    await ctx.set('retries', 0);
    await ctx.set('execution_status', 'initial');

    ctx.logger.info(`Task: ${task_description.slice(0, 100)}...`);
    ctx.logger.info(`Max retries: ${max_retries}`);

    // ── STEP 1: PLANNING ──────────────────────────────────────────────────────
    ctx.logger.info('\n' + '='.repeat(60));
    ctx.logger.info('STEP 1: PLANNING');
    ctx.logger.info('='.repeat(60));

    let plan: { dev_plan: string; test_plan: string };
    try {
      plan = await ctx.step('planning', () =>
        plannerNode(ctx, { task_description }),
      );
    } catch (err) {
      const msg = (err as Error).message;
      ctx.logger.error(`Planning failed: ${msg}`);
      return {
        success: false,
        task: task_description,
        iterations: 0,
        error: `Planning failed: ${msg}`,
      };
    }

    await ctx.set('dev_plan', plan.dev_plan);
    await ctx.set('test_plan', plan.test_plan);
    ctx.logger.info('Planning complete');

    // ── ITERATION LOOP ────────────────────────────────────────────────────────
    for (let iteration = 0; iteration < max_retries; iteration++) {
      ctx.logger.info('\n' + '='.repeat(60));
      ctx.logger.info(`ITERATION ${iteration + 1}/${max_retries}`);
      ctx.logger.info('='.repeat(60));

      const executionStatus =
        (await ctx.get<string>('execution_status')) ?? 'initial';
      const devPlan = (await ctx.get<string>('dev_plan')) ?? '';
      const testPlan = (await ctx.get<string>('test_plan')) ?? '';

      ctx.logger.info('\nSTEP 2: CODE GENERATION');

      let generatedCode: string;
      let generatedTests: string;

      if (iteration === 0) {
        // First iteration: generate tests then code (code sees the tests)
        ctx.logger.info('Generating test suite first');
        const testResult = await ctx.step(
          `test_gen_iter_${iteration}`,
          () =>
            testGeneratorNode(ctx, {
              task_description,
              test_plan: testPlan,
            }),
        );
        generatedTests = testResult.code;

        ctx.logger.info('Generating implementation (guided by tests)');
        const codeResult = await ctx.step(
          `code_gen_iter_${iteration}`,
          () =>
            codeGeneratorNode(ctx, {
              task_description,
              dev_plan: devPlan,
              execution_status: 'initial',
              generated_tests: generatedTests,
            }),
        );
        generatedCode = codeResult.code;
      } else {
        // Subsequent iterations: analyze errors, then fix code
        ctx.logger.info('Fixing code based on test failures');
        ctx.logger.info('\nAnalyzing errors...');

        const prevCode = (await ctx.get<string>('generated_code')) ?? '';
        const prevTests = (await ctx.get<string>('generated_tests')) ?? '';
        const errorLogs = (await ctx.get<string>('error_logs')) ?? '';

        const errorAnalysis = await ctx.step(
          `error_analysis_iter_${iteration}`,
          () =>
            errorAnalyzerNode(ctx, {
              task_description,
              dev_plan: devPlan,
              generated_code: prevCode,
              generated_tests: prevTests,
              error_logs: errorLogs,
            }),
        );

        await ctx.set('error_analysis', errorAnalysis);
        ctx.logger.info(
          `Analysis complete: ${errorAnalysis.analysis_summary.slice(0, 100)}...`,
        );

        const codeResult = await ctx.step(
          `code_fix_iter_${iteration}`,
          () =>
            codeGeneratorNode(ctx, {
              task_description,
              dev_plan: devPlan,
              execution_status: executionStatus,
              generated_code: prevCode,
              generated_tests: prevTests,
              error_logs: errorLogs,
              error_analysis: errorAnalysis,
            }),
        );

        generatedCode = codeResult.code;
        generatedTests = prevTests;
      }

      await ctx.set('generated_code', generatedCode);
      await ctx.set('generated_tests', generatedTests);

      // ── STEP 3: CODE SYNC ──────────────────────────────────────────────────
      ctx.logger.info('\nSTEP 3: CODE SYNC');

      const existingSandboxId = await ctx.get<string>('sandbox_id');
      const syncResult = await ctx.step(
        `code_sync_iter_${iteration}`,
        () =>
          codeSyncNode(ctx, {
            main_code: generatedCode,
            test_code: generatedTests,
            sandbox_id: existingSandboxId,
          }),
      );

      let nextAction: string;
      let status: string;
      let errorLogs: string;
      let sandboxId = existingSandboxId ?? '';

      if (!syncResult.success) {
        ctx.logger.error(`Code sync failed: ${syncResult.message}`);
        await ctx.set('execution_status', 'tests_failed');
        const syncError = syncResult.message ?? 'Code sync failed';
        await ctx.set('error_logs', syncError);
        errorLogs = syncError;
        nextAction = 'retry_code';
        status = 'tests_failed';
      } else {
        sandboxId = syncResult.sandbox_id ?? '';
        await ctx.set('sandbox_id', sandboxId);
        ctx.logger.info(`Synced to sandbox: ${sandboxId}`);

        // ── STEP 4: DEPENDENCY INSTALLATION ───────────────────────────────────
        ctx.logger.info('\nSTEP 4: DEPENDENCY INSTALLATION');
        await ctx.step(
          `install_deps_iter_${iteration}`,
          () =>
            installDepsNode(ctx, {
              main_code: generatedCode,
              sandbox_id: sandboxId,
            }),
        );

        // ── STEP 5: TEST EXECUTION ─────────────────────────────────────────────
        ctx.logger.info('\nSTEP 5: TEST EXECUTION');
        const execResult = await ctx.step(
          `code_exec_iter_${iteration}`,
          () => codeExecutorNode(ctx, { sandbox_id: sandboxId }),
        );

        nextAction = execResult.next_action;
        status = execResult.status;
        errorLogs = execResult.error_logs ?? '';

        await ctx.set('execution_status', status);
        await ctx.set('error_logs', errorLogs);
      }

      // ── STEP 6: DECISION ───────────────────────────────────────────────────
      ctx.logger.info('\nSTEP 6: DECISION');

      if (nextAction === 'success') {
        ctx.logger.info('SUCCESS! All tests passed');
        ctx.logger.info('\nSTEP 7: DOCUMENTATION');

        const finalResult = await ctx.step(
          'final_response',
          () =>
            finalResponseNode(ctx, {
              task_description,
              generated_code: generatedCode,
            }),
        );

        ctx.logger.info('='.repeat(60));
        ctx.logger.info('WORKFLOW COMPLETE - SUCCESS');
        ctx.logger.info('='.repeat(60));

        return {
          success: true,
          task: task_description,
          iterations: iteration + 1,
          code: generatedCode,
          tests: generatedTests,
          sandbox_id: sandboxId,
          documentation: finalResult.markdown_content,
        };
      } else {
        // All non-success outcomes → retry
        const retries = ((await ctx.get<number>('retries')) ?? 0) + 1;
        await ctx.set('retries', retries);

        if (retries >= max_retries) {
          ctx.logger.warn(`Maximum retries (${max_retries}) reached`);
          ctx.logger.info('='.repeat(60));
          ctx.logger.info('WORKFLOW COMPLETE - MAX RETRIES');
          ctx.logger.info('='.repeat(60));
          return {
            success: false,
            task: task_description,
            iterations: iteration + 1,
            error: `Maximum retries (${max_retries}) exhausted`,
            error_logs: errorLogs,
            code: generatedCode,
            tests: generatedTests,
            sandbox_id: sandboxId,
          };
        }

        ctx.logger.warn(`Retrying (${retries}/${max_retries}) — ${status}`);
        // continue to next iteration
      }
    }

    // Should never be reached unless max_retries is 0
    ctx.logger.error('Unexpected workflow termination');
    const finalCode = (await ctx.get<string>('generated_code')) ?? undefined;
    const finalTests = (await ctx.get<string>('generated_tests')) ?? undefined;
    const finalSandboxId = (await ctx.get<string>('sandbox_id')) ?? undefined;

    return {
      success: false,
      task: task_description,
      iterations: max_retries,
      error: 'Workflow terminated unexpectedly',
      code: finalCode,
      tests: finalTests,
      sandbox_id: finalSandboxId,
    };
  },
);

File: go/coding_agent/src/coding_agent/workflow.go
// Coding agent workflow: plan once, then iterate generate -> sync -> install
// -> execute -> (on failure) analyze -> fix, until tests pass or retries run
// out.
//
// Each stage is wrapped in agnt5.Step. Step derives its checkpoint key from
// the step name plus a call-count index, so calling agnt5.Step with the same
// name on every loop iteration is safe — each iteration gets its own
// checkpoint (step:code_generator:0, :1, :2, ...) automatically. On replay
// after a crash, already-completed iterations return their cached result
// immediately instead of re-running.
package coding_agent

import (
	"context"
	"fmt"
	"sort"

	"github.com/agnt5dev/sdk-go/agnt5"
)

const defaultMaxRetries = 15

// Generated tests are written before any code and can be wrong. The error
// analysis may flag such tests, and the workflow corrects them — but at most
// this many times per run, so it cannot keep rewriting tests to escape a
// failing implementation.
const maxTestRepairs = 2

type CodingAgentInput struct {
	TaskDescription string `json:"task_description"`
	MaxRetries      int    `json:"max_retries"`
}

func CodingAgentWorkflow(ctx *agnt5.Context, in CodingAgentInput, model agnt5.LanguageModel, e2b *e2bClient) (WorkflowResult, error) {
	maxRetries := in.MaxRetries
	if maxRetries <= 0 {
		maxRetries = defaultMaxRetries
	}
	taskDescription := in.TaskDescription

	ctx.Logger().Info("INITIATING CODING AGENT WORKFLOW")
	ctx.Logger().Info("Task", "description", truncate(taskDescription, 100), "max_retries", maxRetries)

	ctx.Logger().Info("STEP 1: PLANNING")
	plan, err := agnt5.Step(ctx, "planner", func(context.Context) (Plan, error) {
		return plannerNode(ctx, model, taskDescription)
	})
	if err != nil {
		return failedResult(ctx, taskDescription, 0, "", "", "", fmt.Sprintf("Planning failed: %v", err)), nil
	}
	ctx.Logger().Info("Planning complete")

	var (
		generatedCode, generatedTests string
		sandboxID                     string
		executionStatus               = "initial"
		errorLogs                     string
		testRepairs                   int
		testsRepaired                 []string
	)

	for iteration := 0; iteration < maxRetries; iteration++ {
		ctx.Logger().Info("ITERATION", "n", iteration+1, "of", maxRetries)
		ctx.Logger().Info("STEP 2: CODE GENERATION")

		if iteration == 0 {
			ctx.Logger().Info("Generating test suite first")
			testResult, err := agnt5.Step(ctx, "test_generator", func(context.Context) (GeneratedCode, error) {
				return testGeneratorNode(ctx, model, taskDescription, plan.TestPlan)
			})
			if err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Test generation failed: %v", err)), nil
			}
			generatedTests = testResult.Code

			ctx.Logger().Info("Generating implementation (guided by tests)")
			codeResult, err := agnt5.Step(ctx, "code_generator", func(context.Context) (GeneratedCode, error) {
				return codeGeneratorNode(ctx, model, codeGenInput{
					TaskDescription: taskDescription, DevPlan: plan.DevPlan,
					ExecutionStatus: "initial", GeneratedTests: generatedTests,
				})
			})
			if err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Code generation failed: %v", err)), nil
			}
			generatedCode = codeResult.Code
		} else {
			ctx.Logger().Info("Fixing code based on test failures")
			analysis, err := agnt5.Step(ctx, "error_analyzer", func(context.Context) (ErrorAnalysis, error) {
				return errorAnalyzerNode(ctx, model, taskDescription, plan.DevPlan, generatedCode, generatedTests, errorLogs)
			})
			if err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Error analysis failed: %v", err)), nil
			}
			ctx.Logger().Info("Analysis complete", "summary", truncate(analysis.AnalysisSummary, 100))

			// Some failures may be the tests' fault, not the code's. Correct
			// those tests first, so the fixer is not pushed toward a wrong value.
			analysisForFixer := analysis
			onlyTestFailures := false
			switch {
			case len(analysis.InvalidTests) > 0 && testRepairs < maxTestRepairs:
				ctx.Logger().Info("Correcting tests that contradict the task")
				repair, err := agnt5.Step(ctx, "test_repair", func(context.Context) (TestRepair, error) {
					return testRepairNode(ctx, model, taskDescription, generatedTests, analysis.InvalidTests)
				})
				if err != nil {
					return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Test repair failed: %v", err)), nil
				}
				testRepairs++
				if repair.Accepted {
					generatedTests = repair.Code
					testsRepaired = mergeSorted(testsRepaired, repair.Repaired)
					repaired := map[string]bool{}
					for _, name := range repair.Repaired {
						repaired[name] = true
					}
					var remaining []string
					for _, t := range analysis.FailedTests {
						if !repaired[baseTestName(t)] {
							remaining = append(remaining, t)
						}
					}
					analysisForFixer.FailedTests = remaining
					analysisForFixer.InvalidTests = nil
					onlyTestFailures = len(remaining) == 0
				}
			case len(analysis.InvalidTests) > 0:
				ctx.Logger().Warn("Test repair limit reached; treating failures as code bugs", "limit", maxTestRepairs)
			}

			if onlyTestFailures {
				// Every failure came from a test that is now corrected: the code
				// may already be right, so re-run before changing it.
				ctx.Logger().Info("All failures were in the tests; re-running without changing the code")
			} else {
				codeResult, err := agnt5.Step(ctx, "code_generator", func(context.Context) (GeneratedCode, error) {
					return codeGeneratorNode(ctx, model, codeGenInput{
						TaskDescription: taskDescription, DevPlan: plan.DevPlan,
						ExecutionStatus: executionStatus, GeneratedCode: generatedCode,
						GeneratedTests: generatedTests, ErrorLogs: errorLogs, ErrorAnalysis: &analysisForFixer,
					})
				})
				if err != nil {
					return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Code fix failed: %v", err)), nil
				}
				generatedCode = codeResult.Code
			}
		}

		ctx.Logger().Info("STEP 3: CODE SYNC")
		syncResult, err := agnt5.Step(ctx, "code_sync", func(context.Context) (SyncResult, error) {
			return codeSyncNode(ctx, e2b, generatedCode, generatedTests, sandboxID)
		})
		if err != nil {
			return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Code sync failed: %v", err)), nil
		}

		var nextAction, status string
		if !syncResult.Success {
			ctx.Logger().Error("Code sync failed", "message", syncResult.Message)
			executionStatus = "tests_failed"
			errorLogs = syncResult.Message
			if errorLogs == "" {
				errorLogs = "Code sync failed"
			}
			nextAction, status = "retry_code", "tests_failed"
		} else {
			sandboxID = syncResult.SandboxID
			ctx.Logger().Info("Synced to sandbox", "sandbox_id", sandboxID)

			ctx.Logger().Info("STEP 4: DEPENDENCY INSTALLATION")
			if _, err := agnt5.Step(ctx, "install_deps", func(context.Context) (bool, error) {
				return true, installDepsNode(ctx, e2b, generatedCode, sandboxID)
			}); err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Dependency install failed: %v", err)), nil
			}

			ctx.Logger().Info("STEP 5: TEST EXECUTION")
			execResult, err := agnt5.Step(ctx, "code_executor", func(context.Context) (ExecutionResult, error) {
				return codeExecutorNode(ctx, e2b, sandboxID)
			})
			if err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Test execution failed: %v", err)), nil
			}

			nextAction, status, errorLogs = execResult.NextAction, execResult.Status, execResult.ErrorLogs
			executionStatus = status
		}

		ctx.Logger().Info("STEP 6: DECISION")

		if nextAction == "success" {
			ctx.Logger().Info("SUCCESS! All tests passed")
			if len(testsRepaired) > 0 {
				ctx.Logger().Warn("Passed after correcting generated test(s)", "tests", testsRepaired)
			}
			ctx.Logger().Info("STEP 7: DOCUMENTATION")

			finalResult, err := agnt5.Step(ctx, "final_response", func(context.Context) (FinalResponse, error) {
				return finalResponseNode(ctx, model, taskDescription, generatedCode)
			})
			if err != nil {
				return failedResult(ctx, taskDescription, iteration+1, generatedCode, generatedTests, sandboxID, fmt.Sprintf("Documentation failed: %v", err)), nil
			}

			ctx.Logger().Info("WORKFLOW COMPLETE - SUCCESS")
			return WorkflowResult{
				Success: true, Task: taskDescription, Iterations: iteration + 1,
				Code: generatedCode, Tests: generatedTests, SandboxID: sandboxID,
				Documentation: finalResult.MarkdownContent, TestsRepaired: testsRepaired,
			}, nil
		}

		if iteration == maxRetries-1 {
			ctx.Logger().Warn("Maximum retries reached", "max_retries", maxRetries)
			return WorkflowResult{
				Success: false, Task: taskDescription, Iterations: iteration + 1,
				Error:     fmt.Sprintf("Maximum retries (%d) exhausted", maxRetries),
				ErrorLogs: errorLogs, Code: generatedCode, Tests: generatedTests, SandboxID: sandboxID,
				TestsRepaired: testsRepaired,
			}, nil
		}
		ctx.Logger().Warn("Retrying", "attempt", iteration+2, "of", maxRetries, "status", status)
	}

	return failedResult(ctx, taskDescription, maxRetries, generatedCode, generatedTests, sandboxID, "Workflow terminated unexpectedly"), nil
}

func failedResult(ctx *agnt5.Context, task string, iterations int, code, tests, sandboxID, errMsg string) WorkflowResult {
	ctx.Logger().Error("WORKFLOW COMPLETE - ERROR", "error", errMsg)
	return WorkflowResult{
		Success: false, Task: task, Iterations: iterations,
		Error: errMsg, Code: code, Tests: tests, SandboxID: sandboxID,
	}
}

// mergeSorted returns the union of two name lists, sorted and deduplicated.
func mergeSorted(a, b []string) []string {
	seen := map[string]bool{}
	for _, s := range append(append([]string{}, a...), b...) {
		seen[s] = true
	}
	out := make([]string, 0, len(seen))
	for s := range seen {
		out = append(out, s)
	}
	sort.Strings(out)
	return out
}
